core/ipc/grpc/client/src/main/java/org/opennms/core/ipc/grpc/client/MinionGrpcClient.java

```diff
@@ -0,0 +1,509 @@
+                LOG.info("Registered module {} with gRPC IPC client", rpcModule.getId());
+            }
+        }
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    public void unbind(RpcModule module) throws Exception {
+        if (module != null) {
+            final RpcModule<RpcRequest, RpcResponse> rpcModule = (RpcModule<RpcRequest, RpcResponse>) module;
+            registerdModules.remove(rpcModule.getId());
+            LOG.info("Removing module {} from gRPC IPC client.", rpcModule.getId());
+        }
+    }
+
+    private boolean hasChangedToReadyState() {
+        ConnectivityState prevState = currentChannelState;
+        return !prevState.equals(ConnectivityState.READY) && getChannelState().equals(ConnectivityState.READY);
+    }
+
+    public void shutdown() {
+        requestHandlerExecutor.shutdownNow();
+        blockingSinkMessageScheduler.shutdownNow();
+        registerdModules.clear();
+        if (rpcStream != null) {
+            rpcStream.onCompleted();
+        }
+        channel.shutdown();
+        LOG.info("Minion at location {} with systemId {} stopped", minionIdentity.getLocation(), minionIdentity.getId());
+    }
+
+    public void setBundleContext(BundleContext bundleContext) {
+        this.bundleContext = bundleContext;
+    }
+
+
+    @Override
+    public String getMetricDomain() {
+        return SINK_METRIC_PRODUCER_DOMAIN;
+    }
+
+    @Override
+    public BundleContext getBundleContext() {
+        return bundleContext;
+    }
+
+    @Override
+    public Tracer getTracer() {
+        if (tracerRegistry != null) {
+            return tracerRegistry.getTracer();
+        }
+        return GlobalTracer.get();
+    }
+
+    @Override
+    public MetricRegistry getMetrics() {
+        if (metrics == null) {
+            return new MetricRegistry();
+        }
+        return metrics;
+    }
+
+    public void setMetrics(MetricRegistry metrics) {
+        this.metrics = metrics;
+    }
+
+    public TracerRegistry getTracerRegistry() {
+        return tracerRegistry;
+    }
+
+    public void setTracerRegistry(TracerRegistry tracerRegistry) {
+        this.tracerRegistry = tracerRegistry;
+    }
+
+    ConnectivityState getChannelState() {
+        return currentChannelState = channel.getState(true);
+    }
+
+
+    @Override
+    public <S extends Message, T extends Message> void dispatch(SinkModule<S, T> module, String metadata, T message) {
+
+        try (Logging.MDCCloseable mdc = Logging.withPrefixCloseable(MessageConsumerManager.LOG_PREFIX)) {
+            byte[] sinkMessageContent = module.marshal(message);
+            String messageId = UUID.randomUUID().toString();
+            SinkMessage.Builder sinkMessageBuilder = SinkMessage.newBuilder()
+                    .setMessageId(messageId)
+                    .setLocation(minionIdentity.getLocation())
+                    .setModuleId(module.getId())
+                    .setContent(ByteString.copyFrom(sinkMessageContent));
```

Comment:
Use `ByteString.wrap` instead - bytes wont change.

Reply:
Looks like  ByteString.wrap is not public.  There is no alternative way of other than `copyFrom`

Reply:
Ack. Missed that.